TST: Use placeholders for text in layout tests

QuLogic opened the pull request:
## PR summary

This is an alternative to #29833 which patches the `Text` object to draw a rectangle and not depend on any properties of the actual font. The calculation for the placeholders is somewhat arbitrary, but actually ends up rather close to the size of the original text, so most images don't change a whole lot.

I tested this by changing `matplotlib.testing.set_font_settings_for_testing` from DejaVu Sans to Noto Sans; this would have failed most of the tests before, but none failed now.

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

QuLogic commented on lib/matplotlib/tests/test_constrainedlayout.py:
```diff
@@ -641,7 +644,7 @@ def test_compressed1():
 
     pos = axs[0, 0].get_position()
-    np.testing.assert_allclose(pos.x0, 0.2344, atol=1e-3)
+    np.testing.assert_allclose(pos.x0, 0.2381, atol=1e-2)
```
Note, I found `test_compressed1` to be a bit flaky. Out of 200 repeats, 12 failed on this line or one of the two below where I changed the tolerance. I'm not entirely sure why, as the other tests don't have an issue. If you prefer, I can change the tolerance on all of these lines for consistency.

timhoffm commented on lib/matplotlib/testing/conftest.py:
```diff
@@ -125,3 +125,39 @@ def test_imshow_xarray(xr):
+        """
+        height = fontprop.get_size()
+        width = len(text) * height / 1.618  # Golden ratio for character size.
+        return width, height, 0
```
Texts seem to generally, a somewhat low x-height:

![grafik](https://github.com/user-attachments/assets/a93a5345-623b-47e7-bfc2-5f4c63bfaf0f)

Could this be related to descent=0?

A simple and rougly correct assumtion would be `descent = height // 4`.

QuLogic replied:
It's not really the x-height but the entire height of the text that's smaller. Technically, I believe `get_window_extent` is in pixels, and font size is in points, so I could run this through `renderer.points_to_pixels`, but I found that to be much too large (because this calculation assumes a monospace font while we aren't using one.)

I think `descent=0` is still okay though, as the legend handle is aligned to the centre of the text and the placeholder box. If I set `descent` to anything else, then the line is not longer aligned as such.

jklymak reviewed (approved):
Looks good.  I would never have known to use monkeypatch!

timhoffm commented on lib/matplotlib/testing/conftest.py:
```diff
@@ -125,3 +125,39 @@ def test_imshow_xarray(xr):
+        if self.get_text() == '':
+            return
+        bbox = self.get_window_extent()
+        Rectangle(bbox.p0, bbox.width, bbox.height, color='grey').draw(renderer)
```
```suggestion
        Rectangle(bbox.p0, bbox.width, bbox.height, facecolor='grey').draw(renderer)
```

`color` would also set the edgecolor, which is transparent by default ('none'). The edgecolor induces a line to be drawn around the region and since lines have antialiasing, the edges can become blury. We don't want that. See also https://github.com/matplotlib/matplotlib/issues/29874#issuecomment-2781694549.

QuLogic commented:
I pushed a couple minor tweaks:
1. Changed the box to get its colour from the text. This isn't used in these tests, but could be elsewhere.
2. Removed the edge colour from the box.
3. Added a comment about the bounding box size.
4. Tweaked the importing so that there's only one fixture.

timhoffm commented on lib/matplotlib/testing/conftest.py:
```diff
@@ -125,3 +125,46 @@ def test_imshow_xarray(xr):
+            return
+        bbox = self.get_window_extent()
+        rect = Rectangle(bbox.p0, bbox.width, bbox.height,
+                         facecolor=(self.get_color(), 0.5), edgecolor='none')
```
I'm a bit torn on using transparency here. On the one hand, that correctly catches overlay (but that should be rare). On the other hand, alpha blending is an additional component determining the output image, can that lead to more unstable reference images in the future?

timhoffm replied:
Test images looks a bit redacted now 😆, but I think this is better.

timhoffm commented on lib/matplotlib/testing/conftest.py:
```diff
@@ -125,3 +125,46 @@ def test_imshow_xarray(xr):
+
+
+@pytest.fixture
+def _text_placeholders(monkeypatch):
```
```suggestion
def _text_placeholders(monkeypatch):
    """
    Replace texts with a placeholder rectangle.
    
    The rectangle size only depends on the font size and the number of
    characters. It is thus insensitive to font properties and rendering
    details. This should be used for tests that depend on text geometries
    but not the actual text rendering, e.g. layout tests.
```

timhoffm replied:
Also, do we need to make this private? Little of the other stuff in `matplotlib.testing` is private. I could imagine that this is useful for downstream libraries as well. I don't think we would massively limit us if this is public. Worst case, we could create a new fixture under a different name and deprecate the old one, if we need to make a massively breaking change.

timhoffm reviewed (commented):
This should also be mentioned in 

https://matplotlib.org/devdocs/devel/testing.html#writing-an-image-comparison-test

QuLogic commented:
While updating that doc section, I was also reminded that we should prefer `mpl20` style. I guess I should update all of these, though it would change a bunch of figure sizes.

timhoffm commented:
I would do the mpl20 change in one go. I checked the diffs: The smaller figure sizes are ok. AFAICS no layout information is lost. The only one I'm not sure about is [tight_layout_offsetboxes2.svg](https://github.com/QuLogic/matplotlib/compare/layout-text-placeholders..ltp#diff-d215ea9d93d83f75b90a073952ee8661646713bc0ee13caf13be6f011f6c46e2)

![image](https://github.com/user-attachments/assets/97916a0c-14e5-4701-bc98-9aec6df5f882)

The offset boxes (red/green) and the labels of the right column overlap. But I guess that's just how tightlayout works. It does not account for the actual label sizes.

jklymak commented:
Tight_layout _should_ account for the actual reported label sizes.

timhoffm commented:
So is this an issue due the rectangle not representing the actual text width but rather an approximation?

jklymak commented:
I'm not sure - I dont' see an equivalent test for constrained layout - but maybe it is a draw twice sort of error, where the offset boxes get moved when the axes gets resized, and therefore the margin made for the axes is too small for th new position.  We run constrained layout twice to get around that; usually the position is "closer" after the first time and the new margin is big enough.

timhoffm commented:
Hm @QuLogic maybe the workaround here could be to increase the figure size to not be bothered with this topic right now? 🤯 Not great, but digging into whether this is a tightlayout issue or a box approximation issue also does not feel like a good use of time.

QuLogic commented:
> Tight_layout _should_ account for the actual reported label sizes.

But these are not labels, they are `AnchoredOffsetBox`. Trying it out with constrained layout, it also doesn't seem to interact with the ticks either.

Based on the comment in the test, it appears the intention is to ensure that the boxes are included in the Axes bounding box, but not necessarily have any effect on the ticks labels. Then the _real_ test is the difference between 1 and 2, where the invisible `AnchoredOffsetBox` don't count for the bounding box in the latter, and it looks like that property is preserved.

So I think that a) we can drop this test down to just .png since it's really just checking if a specific artist is included, b) maybe we should remove the tick labels entirely to avoid confusing ourselves, and c) maybe this doesn't need to be an image test at all?

jklymak commented:
Oh I see.  Yes the boxes can overlap the labels in their own axes.  They should not overlap labels in other axes.

QuLogic commented:
Here's my suggestion for removing the image from the test: https://github.com/matplotlib/matplotlib/commit/b6dff32983408a7729ef145a85ffd5e89d38c240

It creates 3 figures:
1. one without any `AnchoredOffsetbox`
2. one with `AnchoredOffsetbox` around each `Axes`
3. one with them invisible `AnchoredOffsetbox`

Then the test is that in figure 2, each `Axes` is smaller than in  figure 1, and no boxes overlap. And then for figure 3, it's that it is the same as figure 1.

If that makes some sense, then I'll rebase this PR on top of that, so we can drop the figures before changing them here.

tacaswell commented:
I am :+1: on moving to the version of that test that @QuLogic has proposed and moving all of these to mpl20 style in this PR.

I think that this is the next PR in the sequence to move to freetype 2.13
